x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/allocation/TrainedModelAllocationClusterService.java

```diff
@@ -0,0 +1,376 @@
+        // Add nodes and keep track of currently routed nodes
+        // TODO: add memory aware node assignments.
+        // Should we indicate a partial allocation somehow if some nodes don't have space?
+        for (Map.Entry<String, TrainedModelAllocation> modelAllocationEntry : previousState.modelAllocations().entrySet()) {
+            for (DiscoveryNode node : currentState.getNodes()) {
+                // Only add the route if the node is NOT shutting down, this would be a weird case of the node
+                // just being added to the cluster and immediately shutting down...
+                if (isNodeShuttingDown(currentState, node.getId()) == false && modelAllocationEntry.getValue().canAllocateToNode(node)) {
+                    nodeHasCapacity(currentState, modelAllocationEntry.getValue().getTaskParams(), node).ifPresentOrElse(
+                        (error) -> builder.addFailedNode(modelAllocationEntry.getKey(), node.getId(), error),
+                        () -> builder.addNode(modelAllocationEntry.getKey(), node.getId())
+                    );
+                }
+            }
+            for (String nodeId : modelAllocationEntry.getValue().getNodeRoutingTable().keySet()) {
+                removedNodeModelLookUp.computeIfAbsent(nodeId, k -> new ArrayList<>()).add(modelAllocationEntry.getKey());
+            }
+        }
+
+        // Remove nodes
+        currentState.getNodes()
+            .forEach(
+                d -> {
+                    // If a node is referenced in the current state, we shouldn't remove the node
+                    // But, if that node that is referenced is shutting down, we should remove the node
+                    if (isNodeShuttingDown(currentState, d.getId()) == false) {
+                        removedNodeModelLookUp.remove(d.getId());
+                    }
+                }
+            );
+        for (Map.Entry<String, List<String>> nodeToModels : removedNodeModelLookUp.entrySet()) {
+            final String nodeId = nodeToModels.getKey();
+            for (String modelId : nodeToModels.getValue()) {
+                builder.removeNode(modelId, nodeId);
+            }
+        }
+        return update(currentState, builder);
+    }
+
+    static boolean shouldAllocateModels(final ClusterChangedEvent event) {
+        // If there are no allocations created at all, there is nothing to update
+        final TrainedModelAllocationMetadata newMetadata = event.state().getMetadata().custom(TrainedModelAllocationMetadata.NAME);
+        if (newMetadata == null) {
+            return false;
+        }
+        boolean masterChanged = event.previousState().nodes().isLocalNodeElectedMaster() == false;
+        if (event.nodesChanged() || masterChanged) {
+            DiscoveryNodes.Delta nodesDelta = event.nodesDelta();
+            for (Map.Entry<String, TrainedModelAllocation> modelAllocationEntry : newMetadata.modelAllocations().entrySet()) {
```

Comment:
As we never access the key of those entries, we could simplify this by making the for loop `for (TrainedModelAllocation allocation : newMetadata.modelAllocations().values())`.